### DISCUSSION TITLE
How to serialize a shape that doesn't have materials as a GLB file ?

### DISCUSSION THREAD

**Discussion by Jarlisco** [IfcOpenShell/IfcOpenShell#4061] [Q&A] [answered]
Hello everyone,

I hope this message finds you well. I'm reaching out for guidance on a specific challenge I'm facing while working with ifcopenshell in Python. My goal is to start from a list of GUIDs, create shapes for those GUIDs, and then attempt to serialize them into GLB files. If a shape cannot be created, I want to throw an error, and if it can, I'd like to serialize it.

Here's a snippet of what I've tried so far:
```
import ifcopenshell
import ifcopenshell.geom
import ifcopenshell.util.shape

ifc_file = ifcopenshell.open('./C8RDC.ifc')
elements_to_convert = ['0ykcHawm52ivsoRSeDyURI', '3AgGEsez10HwSXEKxVOKxX', '2CBjXAg3b75BVWOs9unhsj', '3Sm92wPu5DSw33_zm5pcP3'] # [ifcDoor, ifcWall, ifcSpace, ifcBuildingStorey]

settings = ifcopenshell.geom.settings()
index = 0

for guid in elements_to_convert:
    element = ifc_file.by_guid(guid)
    try:
        shape = ifcopenshell.geom.create_shape(settings, element)
    except Exception as e:
        print(f"An error occurred: {e}")

    sr = ifcopenshell.geom.main.serializers.gltf('result_'+str(index)+'.gltf', settings)
    sr.setFile(ifc_file)
    sr.write(shape)
    sr.finalize()
    index = index + 1
```

My current challenge lies in handling shapes without materials during serialization. If the shape lacks materials, the serializer doesn't seem to work as expected (at least as "I" expect 😄).
Here for example the third guid corresponds to an ifcSpace, from what I understand it can create the shape using the coordinates of the boundingbox but because the shape.geometry.materials is empty it won't serialize (at least 'properly', I can see that the file is created but it is empty). 
Maybe because of that:
`void GltfSerializer::write(const IfcGeom::TriangulationElement* o) {
	if (o->geometry().material_ids().empty()) {
		return;
	}
...`

Can I force some kind of default material onto this shape ?
Also as you can see from my snippet I write a GLTF file, can I make it a GLB directly ? 

I would greatly appreciate any insights, suggestions, or examples on how to handle this situation.

**Comment by aothms** [comment, marked as answer]
I think you can call this a bit of a bug or major inconvenience, but the issue is that when using the serializers you need to use:

```
settings = ifcopenshell.geom.settings(APPLY_DEFAULT_MATERIALS=True)
```

because exactly as you say, the serializers are not programmed to handle elements without materials. This setting generates default materials for elements without.

Previously the serializers weren't exposed to the the python API so this wasn't so much an issue. But probably now there needs to be some better error handling to steer users in the correct direction.

**Reply by Jarlisco** [reply]
Thank you for your response, I am sorry it was this simple and I didn't saw it. This made my code run and finish without errors however the gltf files created are not "working" (import on blender doesn't work for exemple "bad GLTF no asset in json" and when I open it in text editor there seems to be binary data in it). Was my use of serializers even correct in the snippet I posted earlier ?

I'll mark your response as the Answer since it truly is what I was looking for!

**Reply by aothms** [reply]
I had another look. I'd propose the following three changes:

WELD_VERTICES=False is better suited for OpenGL type rendering.

- `settings = ifcopenshell.geom.settings(APPLY_DEFAULT_MATERIALS=True, WELD_VERTICES=False)`
- `sr.writeHeader()` before sr.write()
- `+'.glb'` use a glb filename extension since the result is a **binary** gltf.
